packages/frontend/src/components/chat.tsx

```diff
@@ -45,37 +52,108 @@ function Message({ msg, from, peerId }: MessageProps) {
+        case CHAT_TOPIC: {
+          chatMessageCB(evt, topic, data)
+          break
+        }
+        case CHAT_FILE_TOPIC: {
+          chatFileMessageCB(evt, topic, data)
+          break
+        }
+        default: {
+          console.log(`Unexpected gossipsub topic: ${topic}`)
+          break
+        }
+      }
+    }
+
+    const chatMessageCB = (evt: CustomEvent<Message>, topic: string, data: Uint8Array) => {
       const msg = new TextDecoder().decode(data)
       console.log(`${topic}: ${msg}`)
 
       // Append signed messages, otherwise discard
       if (evt.detail.type === 'signed') {
-        setMessageHistory([...messageHistory, { msg, from: 'other', peerId: evt.detail.from.toString() }])
+        setMessageHistory([...messageHistory, { msg, fileObjectUrl: undefined, from: 'other', peerId: evt.detail.from.toString() }])
+      }
+    }
+
+    const chatFileMessageCB = async (evt: CustomEvent<Message>, topic: string, data: Uint8Array) => {
+      const fileId = new TextDecoder().decode(data)
+      console.log(`${topic}: ${fileId}`)
+
+      // if the message isn't signed, discard it.
+      if (evt.detail.type !== 'signed') {
+        return
       }
+      const senderPeerId = evt.detail.from;
+      console.log(`${topic}: ${fileId} from ${senderPeerId}`)
+
+      const stream = await libp2p.dialProtocol(senderPeerId, FILE_EXCHANGE_PROTOCOL)
+      await pipe(
+        [uint8ArrayFromString(fileId)],
+        (source) => lp.encode(source),
+        stream,
+        (source) => lp.decode(source),
+        async function(source) {
+          for await (const data of source) {
+            const body: Uint8Array = data.subarray()
+            console.log(`request_response: response received: size:${body.length}`)
+
+            const msg: ChatMessage = {
+              msg: newChatFileMessage(fileId, body),
+              fileObjectUrl: window.URL.createObjectURL(new Blob([body])),
+              from: 'other',
+              peerId: senderPeerId.toString(),
+            }
+            setMessageHistory([...messageHistory, msg])
+          }
+        }
+      )
     }
 
     libp2p.services.pubsub.addEventListener('message', messageCB)
 
+    libp2p.handle(FILE_EXCHANGE_PROTOCOL, ({ stream }) => {
+      pipe(
+        stream.source,
+        (source) => lp.decode(source),
+        (source) => map(source, async (msg) => {
+          const fileId = uint8ArrayToString(msg.subarray())
+          console.log(`request_response: request received: fileId:${fileId}, source:${stream.source}`)
+          const file = files.get(fileId)!
+          return file.body
+        }),
+        (source) => lp.encode(source),
+        stream.sink,
+      )
+    })
+
     return () => {
       // Cleanup handlers 👇
       // libp2p.pubsub.unsubscribe(CHAT_TOPIC)
+      // libp2p.pubsub.unsubscribe(CHAT_FILE_TOPIC)
       libp2p.services.pubsub.removeEventListener('message', messageCB)
+      libp2p.unhandle(FILE_EXCHANGE_PROTOCOL)
```

Comment:
This is an [async method](https://github.com/libp2p/js-libp2p/blob/5315f7bc720e8e45ed7bdafa8397b6a1acc2fe4b/packages/libp2p/src/libp2p.ts#L363) so it should be awaited.

Reply:
Thank you. It seems that it's not simple to make this cleanup function as async in the useEffect hook (maybe because I haven't had no experience in React so far :)) I'll find a way.

Reply:
8c3f766. I found a way. I checked that the `unhandle` is called successfully.